src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -31,18 +31,36 @@ public static class ActivatorUtilities
             [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type instanceType,
             params object[] parameters)
         {
-            int bestLength = -1;
-            bool seenPreferred = false;
+            if (provider == null)
+            {
+                throw new ArgumentNullException(nameof(provider));
+            }
 
-            ConstructorMatcher bestMatcher = default;
+            if (instanceType.IsAbstract)
+            {
+                throw new InvalidOperationException(SR.CannotCreateAbstractClasses);
+            }
 
-            if (!instanceType.IsAbstract)
+            IServiceProviderIsService? serviceProviderIsService = provider.GetService<IServiceProviderIsService>();
+            // if container supports using IServiceProviderIsService, we try to find the longest ctor that
+            // (a) matches all parameters given to CreateInstance
+            // (b) matches the rest of ctor arguments as either a parameter with a default value or as a service registered
+            // if no such match is found we fallback to the same logic used by CreateFactory which would only allow creating an
+            // instance if all parameters given to CreateInstance only match with a single ctor
+            if (serviceProviderIsService != null &&
+                serviceProviderIsService.IsService(typeof(IServiceProviderIsService)))
```

Comment:
I'm not sure this check is necessary. We should just need to check if `serviceProviderIsService` is null.

Reply:
cc @halter73 do you have a scenario in which having both checks here would detect properly if a container is misconfigured?

Reply:
I agree we should remove the `IsService(typeof(IServiceProviderIsService))` check. I don't think we can detect a misconfigured container ahead of time. It's just that if `IServiceProviderIsService` returns false so much that no constructors match, we should at least try the constructor `CreateFactory` would have called in the same situation assuming it would be unambiguous.

Reply:
Removed the redundant check.

As for setting up a test for this case, I'm open to suggestions on how to set it up.